File: app/parsers/gis.py
from app.parsers.base import BaseParser
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
import time


class GisParser(BaseParser):
    def build_driver_options(self):
        options = Options()
        options.add_argument("--disable-gpu")
        options.add_argument("--no-sandbox")
        options.add_argument("--disable-dev-shm-usage")
        options.add_argument("--window-size=1920,1080")
        return options

    def extra_fields(self):
        return {
            # Добавляет дополнительные поля в парсинг
        }

    def getName(self):
        try:
            el = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located(
                    ((By.TAG_NAME, "h1"))
                ) 
            )
            return el.text
        except:
            return None
    

    def getRating(self):
        try:
            el = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located(
                    (By.CLASS_NAME, "_y10azs")
                )
            )
            return el.text
        except:
            return None
    

    def getCountReviews(self):
        try:
            el = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located(
                    (By.CLASS_NAME, "_jspzdm")
                )
            )
            return el.text
        except:
            return None


    def openReviewsPage(self):
        try:
            tab = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located(
                    (By.XPATH, "//a[contains(@href,'tab/reviews')]")
                )
            )

            self.driver.execute_script("arguments[0].scrollIntoView(true);", tab)
            self.driver.execute_script("arguments[0].click();", tab)

            WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.CLASS_NAME, "_1k5soqfl"))
            )
            return True

        except Exception as e:
            print("ERROR openReviewsPage:", e)
            return False
        
    def getReviews(self, limit=10):
        if not self.openReviewsPage():
            return []
    
        loaded = 0
        reviews = []
    
        for _ in range(6):  
            self.driver.execute_script("window.scrollBy(0, 600)")
            time.sleep(0.4)
    
            blocks = self.driver.find_elements(By.CLASS_NAME, "_1k5soqfl")
    
            for block in blocks:
                if loaded >= limit:
                    return reviews
    
                try:
                    name_el = block.find_elements(By.CLASS_NAME, "_16s5yj36")
                    if name_el:
                        name = name_el[0].get_attribute("textContent").strip()
                    else:
                        name = None
                except:
                    name = None
                
                try:
                    text_el = block.find_elements(By.CLASS_NAME, "_1msln3t")
                    if text_el:
                        text = text_el[0].get_attribute("textContent").strip()
                    else:
                        text = None
                except:
                    text = None

                try:
                    rating_block = block.find_elements(By.CLASS_NAME, "_1fkin5c")
                    if rating_block:
                        raiting = len(rating_block[0].find_elements(By.TAG_NAME, "span"))
                    else:
                        raiting = None
                except:
                    raiting = None

                try:
                    avatar_el = block.find_element(By.CLASS_NAME, "_1dk5lq4")
                    style = avatar_el.get_attribute("style")
                    avatar = None
                    if style and "background-image" in style:
                        avatar = style.split("url(")[1].split(")")[0]
                        avatar = avatar.replace('"', "").replace("'", "")
                except:
                    avatar = None
    
                reviews.append({
                    "name": name,
                    "text": text,
                    "raiting": raiting,
                    "avatar": avatar,
                })
    
                loaded += 1
        return reviews

File: src/parsers/gis.js
import { By, until } from "selenium-webdriver";
import { withDriver } from "../selenium.js";

async function tryText(driver, locator, timeoutMs = 10000) {
    try {
        const el = await driver.wait(until.elementLocated(locator), timeoutMs);
        return (await el.getText()) || null;
    } catch {
        return null;
    }
}

async function openReviewsPage(driver) {
    try {
        const tab = await driver.wait(until.elementLocated(By.xpath("//a[contains(@href,'tab/reviews')]")), 10000);
        await driver.executeScript("arguments[0].scrollIntoView(true);", tab);
        await driver.executeScript("arguments[0].click();", tab);
        await driver.wait(until.elementLocated(By.className("_1k5soqfl")), 10000);
        return true;
    } catch (e) {
        console.error("ERROR openReviewsPage:", e?.message ?? e);
        return false;
    }
}

export async function parse2gis(url, profileId) {
    return await withDriver(profileId, async (driver) => {
        await driver.get(url);
        
        const name = await tryText(driver, By.css("h1"));
        const rating = await tryText(driver, By.className("_y10azs"));
        const countReviews = await tryText(driver, By.className("_jspzdm"));
        
        const ok = await openReviewsPage(driver);
        if (!ok) return { name, rating, count_reviews: countReviews, reviews: [] };
        
        const reviews = [];
        let loaded = 0;

        for (let i = 0; i < 6 && loaded < 10; i++) {
            await driver.executeScript("window.scrollBy(0, 600)");
            await driver.sleep(400);

            const blocks = await driver.findElements(By.className("_1k5soqfl"));
            for (const block of blocks) {
                if (loaded >= 10) break;

                let reviewerName = null;
                let text = null;
                let avatar = null;
                let raiting = null;

                try {
                  const els = await block.findElements(By.className("_16s5yj36"));
                  reviewerName = els[0] ? (await els[0].getAttribute("textContent")).trim() : null;
                } catch {}

                try {
                  const els = await block.findElements(By.className("_1msln3t"));
                  text = els[0] ? (await els[0].getAttribute("textContent")).trim() : null;
                } catch {}

                try {
                  const ratingBlocks = await block.findElements(By.className("_1fkin5c"));
                  if (ratingBlocks[0]) {
                    const spans = await ratingBlocks[0].findElements(By.css("span"));
                    raiting = spans.length;
                  }
                } catch {}

                try {
                  const avatarEl = await block.findElement(By.className("_1dk5lq4"));
                  const style = await avatarEl.getAttribute("style");
                  if (style && style.includes("url(")) {
                    avatar = style.split("url(")[1].split(")")[0].replace(/["']/g, "");
                  }
                } catch {}

                reviews.push({ name: reviewerName, text, raiting, avatar });
                loaded++;
            }
        }

        return { name, rating, count_reviews: countReviews, reviews };
    });
}
